pkgs/build-support/node/fetch-npm-deps/src/parse/mod.rs

```diff
@@ -0,0 +1,364 @@
+        let mut s = url.path_segments()?;
+
+        match url.host_str()? {
+            "github.com" => {
+                let user = s.next()?;
+                let mut project = s.next()?;
+                let typ = s.next();
+                let mut commit = s.next();
+
+                if typ.is_none() {
+                    commit = url.fragment();
+                } else if typ.is_some() && typ != Some("tree") {
+                    return None;
+                }
+
+                if project.ends_with(".git") {
+                    project = project.strip_suffix(".git")?;
+                }
+
+                let commit = commit.unwrap();
+
+                Some(
+                    Url::parse(&format!(
+                        "https://codeload.github.com/{user}/{project}/tar.gz/{commit}"
+                    ))
+                    .ok()?,
+                )
+            }
+            "bitbucket.org" => {
+                let user = s.next()?;
+                let mut project = s.next()?;
+                let aux = s.next();
+
+                if aux == Some("get") {
+                    return None;
+                }
+
+                if project.ends_with(".git") {
+                    project = project.strip_suffix(".git")?;
+                }
+
+                let commit = url.fragment()?;
+
+                Some(
+                    Url::parse(&format!(
+                        "https://bitbucket.org/{user}/{project}/get/{commit}.tar.gz"
+                    ))
+                    .ok()?,
+                )
+            }
+            "gitlab.com" => {
+                let path = &url.path()[1..];
+
+                if path.contains("/~/") || path.contains("/archive.tar.gz") {
+                    return None;
+                }
+
+                let user = s.next()?;
+                let mut project = s.next()?;
+
+                if project.ends_with(".git") {
+                    project = project.strip_suffix(".git")?;
+                }
+
+                let commit = url.fragment()?;
+
+                Some(
+                    Url::parse(&format!(
+                    "https://gitlab.com/{user}/{project}/repository/archive.tar.gz?ref={commit}"
+                ))
+                    .ok()?,
+                )
+            }
+            "git.sr.ht" => {
+                let user = s.next()?;
+                let mut project = s.next()?;
+                let aux = s.next();
+
+                if aux == Some("archive") {
+                    return None;
+                }
+
+                if project.ends_with(".git") {
+                    project = project.strip_suffix(".git")?;
+                }
+
+                let commit = url.fragment()?;
+
+                Some(
+                    Url::parse(&format!(
+                        "https://git.sr.ht/{user}/{project}/archive/{commit}.tar.gz"
+                    ))
+                    .ok()?,
+                )
+            }
+            _ => None,
```

Comment:
We need a better error message here, because when a git URL is used that is not from a hosted git service, this `None` causes the calling code to assume it is a registry dep rather than git dep and fails saying some expected fields are missing (which are absent since it's not a registry dep)

Reply:
The main issue with this is then deciding what's a registry dep and what's not. I fixed this by no longer checking if something is a Git dep if their scheme is `http` or `https`, as `hosted-git-info` does. (Do we even care about this case?)

Reply:
Looks good now!

If you're asking whether we care to support arbitrary git URLs, I don't think so until someone brings it up as problem. I feel it's unlikely to ever come up, so there's no real reason to implement it until there is an actual need

Reply:
> If you're asking whether we care to support arbitrary git URLs, I don't think so until someone brings it up as problem. I feel it's unlikely to ever come up, so there's no real reason to implement it until there is an actual need

It's impossible to do so with how npm implements it.

What I mean is if we should support URLs in the lockfile that look like `https://github.com/foo/bar`, when, as far as I know, they should always get resolved to `git+` protocols. (Look at the protocols we filter for in the old and new impl.)

Reply:
Ah that's what you meant. Well they should always be resolved to `git+[proto]` so I don't think we need to worry about that